src/main/java/net/neoforged/neoforge/transfer/handlers/resources/IResourceHandler.java

```diff
@@ -0,0 +1,230 @@
+/*
+ * Copyright (c) NeoForged and contributors
+ * SPDX-License-Identifier: LGPL-2.1-only
+ */
+
+package net.neoforged.neoforge.transfer.handlers.resources;
+
+import com.google.common.primitives.Ints;
+import net.neoforged.neoforge.transfer.TransferPreconditions;
+import net.neoforged.neoforge.transfer.resources.IResource;
+import net.neoforged.neoforge.transfer.transaction.SnapshotJournal;
+import net.neoforged.neoforge.transfer.transaction.Transaction;
+import net.neoforged.neoforge.transfer.transaction.TransactionContext;
+import org.jetbrains.annotations.ApiStatus;
+
+/**
+ * A generic handler for the transfer and storage of {@link IResource resources} whether it be inserting, extracting, querying some value, etc.
+ *
+ * <h2>Indices</h2>
+ * <p>A resource handler is organized into indices, which are addressed using an int between {@code 0} and {@code size() - 1}.
+ * <p>Out-of-bounds access using methods that accept an {@code int slot} will usually throw an exception,
+ * so only indices between 0 (included) and the size (excluded) should be used.
+ * If a storage has a dynamic size, it should be lenient to accommodate for callers
+ * holding onto a previously returned size.
+ *
+ * @param <T> The type of resource this handler manages.
+ */
+public interface IResourceHandler<T extends IResource> {
```

Comment:
To me most of these methods should be documented with the case that they can throw certain exceptions in certain cases.

Examples would be `IndexOutOfRangeException` etc.

Reply:
See this section of the class javadoc, which also addresses the problem of dynamic handlers (i.e. they should be lenient):
```
 * <h2>Indices</h2>
 * <p>A resource handler is organized into indices, which are addressed using an int between {@code 0} and {@code size() - 1}.
 * <p>Out-of-bounds access using methods that accept an {@code int slot} will usually throw an exception,
 * so only indices between 0 (included) and the size (excluded) should be used.
 * If a storage has a dynamic size, it should be lenient to accommodate for callers
 * holding onto a previously returned size.
```